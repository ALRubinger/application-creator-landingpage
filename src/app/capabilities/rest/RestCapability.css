#rest-capability button.http{border-radius:0}#rest-capability input.http-param{border-radius:0;height:44px}#rest-capability input#http-api-path-input{background-color:#f1f1f1;border:1px solid #bbb;height:44px;border-radius:0;width:300px;color:black}
