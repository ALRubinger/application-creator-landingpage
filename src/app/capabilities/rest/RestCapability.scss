
#rest-capability {
  button.http {
    border-radius: 0;
  }

  input.http-param {
    border-radius: 0;
    height: 44px;
  }

  input#http-api-path-input{
    background-color: #f1f1f1;
    border: 1px solid #bbb;
    height: 44px;
    border-radius: 0;
    width: 300px;
    color: black;
  }
}