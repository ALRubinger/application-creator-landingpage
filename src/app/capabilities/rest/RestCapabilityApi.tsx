import {HttpApi} from '../../../shared/utils/HttpApi';
import {ApiFactoryFunction} from '../connectCapability';

export interface RestCapabilityApi {
  getGreetingAbsoluteUrl(name: string): string;
  doGetGreeting(name: string): Promise<{content: string, time: number}>;
}

function buildGreetingPath(name: string) {
  return name.length === 0 ? '/greeting' : `/greeting?name=${encodeURIComponent(name)}`;
}

class HttpRestCapabilityApi implements RestCapabilityApi {

  constructor(private readonly httpApi: HttpApi){}

  public async doGetGreeting(name: string): Promise<{content: string, time: number}> {
    const greetingPath = buildGreetingPath(name);
    const r = await this.httpApi.get<{content: string;}>(greetingPath);
    return ({ content: r.content, time: Date.now() });
  }

  public getGreetingAbsoluteUrl(name: string): string {
    return this.httpApi.getApiAbsoluteUrl(buildGreetingPath(name));
  }
}

export class MockRestCapabilityApi implements RestCapabilityApi {
  public async doGetGreeting(name: string): Promise<{content: string, time: number}> {
    return { content: 'Hello ' + (name || 'World') + '!', time: Date.now() };
  }

  public getGreetingAbsoluteUrl(name: string): string {
    return `http://mocked.io/api${buildGreetingPath(name)}`;
  }
}

export const restCapabilityApiFactory:ApiFactoryFunction<RestCapabilityApi> = ({ httpApi, isMockMode }) =>
  isMockMode ? new MockRestCapabilityApi() : new HttpRestCapabilityApi(httpApi);

