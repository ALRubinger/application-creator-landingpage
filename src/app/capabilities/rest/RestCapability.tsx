import * as React from 'react';
import Capability from '../Capability';
import ShellCommand from '../../../shared/components/ShellCommand';
import * as moment from 'moment';

import './RestCapability.css';
import {Button, Grid, GridItem, Split, SplitItem, TextInput} from '@patternfly/react-core';
import {Console} from '../../../shared/components/Console';
import {RestCapabilityApi} from './RestCapabilityApi';
import capabilitiesConfig from '../../config/capabilitiesConfig';

interface RestCapabilityProps {
  apiService: RestCapabilityApi;
}

interface RestCapabilityState {
  consoleContent: string;
  results: Array<{content: string, time: number}>;
  params: {
    [name: string]: string;
  }
}

export default class RestCapability extends React.Component<RestCapabilityProps, RestCapabilityState> {

  constructor(props) {
    super(props);

    this.state = {
      results: [],
      consoleContent: '',
      params: {
        name: '',
      },
    };
  }

  public render() {
    return (
      <Capability module="rest">
        <Capability.Title>{capabilitiesConfig.rest.name}</Capability.Title>
        <Capability.Body>
          <Grid>
            <GridItem span={12}>
              <Split>
                <SplitItem isMain={false}>
                  <TextInput id="http-api-path-input" isDisabled={true} value={`/greetings/`}/>
                </SplitItem>
                <SplitItem isMain={true}>
                  <TextInput id="http-api-name-input"
                             value={this.state.params.name}
                             onChange={this.handleInputChange}
                             name="name" placeholder="World" className="http-param"/>
                </SplitItem>
                <SplitItem isMain={false}>
                  <Button
                    className={'http'}
                    onClick={this.execGet}
                    title="Execute GET Request"
                  >GET</Button>
                </SplitItem>
              </Split>
              <Console id="rest-content" content={this.state.consoleContent}/>
            </GridItem>
            <GridItem span={12}>
              <p>
                You may test this directly by making an <samp>HTTP GET</samp> request using this
                application's URL as root. For instance, try with the <samp>cURL</samp> tool:
              </p>
              <ShellCommand command={`curl ${this.getGreetingsUrl()}`}/>
            </GridItem>
          </Grid>
        </Capability.Body>
      </Capability>
    );
  }

  private execGet = (values) => {
    this.props.apiService.doGetGreeting(this.state.params.name).then(result => {
      this.setState({
        results: [...this.state.results, result],
      });
      this.logToConsole(result);
    });
  }


  private handleInputChange = (_, event) => {
    const target = event.target;
    const value = target.type === 'checkbox' ? target.checked : target.value;
    const name = target.name;

    this.setState({
      params: {
        ...this.state.params,
        [name]: value,
      }
    });
  }

  private logToConsole(result: {content: string, time: number}) {
    const url = this.getGreetingsUrl();
    this.setState({
      consoleContent: `> ${moment(result.time).format('LTS')} GET ${url}: ${result.content}\n${this.state.consoleContent}`
    });
  }

  private getGreetingsUrl() {
    return this.props.apiService.getGreetingAbsoluteUrl(this.state.params.name);
  }
}