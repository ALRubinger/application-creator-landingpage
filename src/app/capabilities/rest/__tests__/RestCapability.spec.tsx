import * as React from 'react';
import {shallow} from 'enzyme';
import RestCapability from '../RestCapability';
import {MockRestCapabilityApi} from '../RestCapabilityApi';


describe('<RestCapability />', () => {
  it('check that initial render is correct', () => {
    const component = shallow(<RestCapability apiService={new MockRestCapabilityApi()}/>);
    expect(component).toMatchSnapshot();
  });

  it('check that a click on the GET button adds a Hello World message in the console', async () => {
    const apiService = new MockRestCapabilityApi();
    const doGetGreetingSpy = jest.spyOn(apiService, 'doGetGreeting');
    const component = shallow(<RestCapability apiService={apiService}/>);
    const result = { content: 'Hello World!', time: 1542793377 };
    const promise = Promise.resolve(result);
    doGetGreetingSpy.mockReturnValue(promise);
    component.find('Button.http').simulate('click');
    expect(doGetGreetingSpy).toHaveBeenCalledWith('');
    await promise;
    component.update();
    expect(component.state('results')).toHaveLength(1);
    expect(component.state('results')[0]).toBe(result);
    expect(component).toMatchSnapshot();
  });

  it('check that after typing a name, a click on the GET button adds a Hello John message in the console', async () => {
    const apiService = new MockRestCapabilityApi();
    const doGetGreetingSpy = jest.spyOn(apiService, 'doGetGreeting');
    const component = shallow(<RestCapability apiService={apiService}/>);
    component.find('TextInput#http-api-name-input').simulate('change', false, { target: { name: 'name', value: 'John'} });
    const result = { content: 'Hello John!', time: 1542793377 };
    const promise = Promise.resolve(result);
    doGetGreetingSpy.mockReturnValue(promise);
    component.find('Button.http').simulate('click');
    expect(doGetGreetingSpy).toHaveBeenCalledWith('John');
    await promise;
    component.update();
    expect(component.state('results')).toHaveLength(1);
    expect(component.state('results')[0]).toBe(result);
    expect(component).toMatchSnapshot();
  });

});

