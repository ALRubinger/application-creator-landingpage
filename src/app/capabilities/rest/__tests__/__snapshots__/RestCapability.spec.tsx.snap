// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RestCapability /> check that a click on the GET button adds a Hello World message in the console 1`] = `
<Capability
  module="rest"
>
  <Component>
    Http Api
  </Component>
  <Component>
    <Grid
      className=""
      gutter={null}
      lg={null}
      md={null}
      sm={null}
      span={null}
      xl={null}
    >
      <GridItem
        className=""
        lg={null}
        lgOffset={null}
        lgRowSpan={null}
        md={null}
        mdOffset={null}
        mdRowSpan={null}
        offset={null}
        rowSpan={null}
        sm={null}
        smOffset={null}
        smRowSpan={null}
        span={12}
        xl={null}
        xlOffset={null}
        xlRowSpan={null}
      >
        <Split
          className=""
          component="div"
          gutter={null}
        >
          <SplitItem
            className=""
            isMain={false}
          >
            <TextInput
              aria-label={null}
              className=""
              id="http-api-path-input"
              isAlt={false}
              isDisabled={true}
              isReadOnly={false}
              isRequired={false}
              isValid={true}
              onChange={[Function]}
              type="text"
              value="/greetings/"
            />
          </SplitItem>
          <SplitItem
            className=""
            isMain={true}
          >
            <TextInput
              aria-label={null}
              className="http-param"
              id="http-api-name-input"
              isAlt={false}
              isDisabled={false}
              isReadOnly={false}
              isRequired={false}
              isValid={true}
              name="name"
              onChange={[Function]}
              placeholder="World"
              type="text"
              value=""
            />
          </SplitItem>
          <SplitItem
            className=""
            isMain={false}
          >
            <Button
              aria-label={null}
              className="http"
              component="button"
              isActive={false}
              isBlock={false}
              isDisabled={false}
              isFocus={false}
              isHover={false}
              onClick={[Function]}
              title="Execute GET Request"
              type="button"
              variant="primary"
            >
              GET
            </Button>
          </SplitItem>
        </Split>
        <Console
          content="> 9:33:13 PM GET http://mocked.io/api/greeting: Hello World!
"
          id="rest-content"
        />
      </GridItem>
      <GridItem
        className=""
        lg={null}
        lgOffset={null}
        lgRowSpan={null}
        md={null}
        mdOffset={null}
        mdRowSpan={null}
        offset={null}
        rowSpan={null}
        sm={null}
        smOffset={null}
        smRowSpan={null}
        span={12}
        xl={null}
        xlOffset={null}
        xlRowSpan={null}
      >
        <p>
          You may test this directly by making an 
          <samp>
            HTTP GET
          </samp>
           request using this application's URL as root. For instance, try with the 
          <samp>
            cURL
          </samp>
           tool:
        </p>
        <ShellCommand
          command="curl http://mocked.io/api/greeting"
        />
      </GridItem>
    </Grid>
  </Component>
</Capability>
`;

exports[`<RestCapability /> check that after typing a name, a click on the GET button adds a Hello John message in the console 1`] = `
<Capability
  module="rest"
>
  <Component>
    Http Api
  </Component>
  <Component>
    <Grid
      className=""
      gutter={null}
      lg={null}
      md={null}
      sm={null}
      span={null}
      xl={null}
    >
      <GridItem
        className=""
        lg={null}
        lgOffset={null}
        lgRowSpan={null}
        md={null}
        mdOffset={null}
        mdRowSpan={null}
        offset={null}
        rowSpan={null}
        sm={null}
        smOffset={null}
        smRowSpan={null}
        span={12}
        xl={null}
        xlOffset={null}
        xlRowSpan={null}
      >
        <Split
          className=""
          component="div"
          gutter={null}
        >
          <SplitItem
            className=""
            isMain={false}
          >
            <TextInput
              aria-label={null}
              className=""
              id="http-api-path-input"
              isAlt={false}
              isDisabled={true}
              isReadOnly={false}
              isRequired={false}
              isValid={true}
              onChange={[Function]}
              type="text"
              value="/greetings/"
            />
          </SplitItem>
          <SplitItem
            className=""
            isMain={true}
          >
            <TextInput
              aria-label={null}
              className="http-param"
              id="http-api-name-input"
              isAlt={false}
              isDisabled={false}
              isReadOnly={false}
              isRequired={false}
              isValid={true}
              name="name"
              onChange={[Function]}
              placeholder="World"
              type="text"
              value="John"
            />
          </SplitItem>
          <SplitItem
            className=""
            isMain={false}
          >
            <Button
              aria-label={null}
              className="http"
              component="button"
              isActive={false}
              isBlock={false}
              isDisabled={false}
              isFocus={false}
              isHover={false}
              onClick={[Function]}
              title="Execute GET Request"
              type="button"
              variant="primary"
            >
              GET
            </Button>
          </SplitItem>
        </Split>
        <Console
          content="> 9:33:13 PM GET http://mocked.io/api/greeting?name=John: Hello John!
"
          id="rest-content"
        />
      </GridItem>
      <GridItem
        className=""
        lg={null}
        lgOffset={null}
        lgRowSpan={null}
        md={null}
        mdOffset={null}
        mdRowSpan={null}
        offset={null}
        rowSpan={null}
        sm={null}
        smOffset={null}
        smRowSpan={null}
        span={12}
        xl={null}
        xlOffset={null}
        xlRowSpan={null}
      >
        <p>
          You may test this directly by making an 
          <samp>
            HTTP GET
          </samp>
           request using this application's URL as root. For instance, try with the 
          <samp>
            cURL
          </samp>
           tool:
        </p>
        <ShellCommand
          command="curl http://mocked.io/api/greeting?name=John"
        />
      </GridItem>
    </Grid>
  </Component>
</Capability>
`;

exports[`<RestCapability /> check that initial render is correct 1`] = `
<Capability
  module="rest"
>
  <Component>
    Http Api
  </Component>
  <Component>
    <Grid
      className=""
      gutter={null}
      lg={null}
      md={null}
      sm={null}
      span={null}
      xl={null}
    >
      <GridItem
        className=""
        lg={null}
        lgOffset={null}
        lgRowSpan={null}
        md={null}
        mdOffset={null}
        mdRowSpan={null}
        offset={null}
        rowSpan={null}
        sm={null}
        smOffset={null}
        smRowSpan={null}
        span={12}
        xl={null}
        xlOffset={null}
        xlRowSpan={null}
      >
        <Split
          className=""
          component="div"
          gutter={null}
        >
          <SplitItem
            className=""
            isMain={false}
          >
            <TextInput
              aria-label={null}
              className=""
              id="http-api-path-input"
              isAlt={false}
              isDisabled={true}
              isReadOnly={false}
              isRequired={false}
              isValid={true}
              onChange={[Function]}
              type="text"
              value="/greetings/"
            />
          </SplitItem>
          <SplitItem
            className=""
            isMain={true}
          >
            <TextInput
              aria-label={null}
              className="http-param"
              id="http-api-name-input"
              isAlt={false}
              isDisabled={false}
              isReadOnly={false}
              isRequired={false}
              isValid={true}
              name="name"
              onChange={[Function]}
              placeholder="World"
              type="text"
              value=""
            />
          </SplitItem>
          <SplitItem
            className=""
            isMain={false}
          >
            <Button
              aria-label={null}
              className="http"
              component="button"
              isActive={false}
              isBlock={false}
              isDisabled={false}
              isFocus={false}
              isHover={false}
              onClick={[Function]}
              title="Execute GET Request"
              type="button"
              variant="primary"
            >
              GET
            </Button>
          </SplitItem>
        </Split>
        <Console
          content=""
          id="rest-content"
        />
      </GridItem>
      <GridItem
        className=""
        lg={null}
        lgOffset={null}
        lgRowSpan={null}
        md={null}
        mdOffset={null}
        mdRowSpan={null}
        offset={null}
        rowSpan={null}
        sm={null}
        smOffset={null}
        smRowSpan={null}
        span={12}
        xl={null}
        xlOffset={null}
        xlRowSpan={null}
      >
        <p>
          You may test this directly by making an 
          <samp>
            HTTP GET
          </samp>
           request using this application's URL as root. For instance, try with the 
          <samp>
            cURL
          </samp>
           tool:
        </p>
        <ShellCommand
          command="curl http://mocked.io/api/greeting"
        />
      </GridItem>
    </Grid>
  </Component>
</Capability>
`;
